Add forc feature for overriding packages in the package graph, akin to cargo's [patch] feature

kayagokalp opened the pull request:
~This is still WIP, opening up for visibility.~

- [x] Support for patching while creating the `BuildPlan` from the manifest file.
- [x] Documentation.
- [x] Support for patching while creating the `BuildPlan` from the lock file.
- [x] Address #1869

## About this PR
implements and closes #1784.

- Introduces `[package]` table to the manifest file.
- While creating a `source` from a `dependency` check for patches and applies them if there are any.

### Testing

**Patching `std` with a different branch**

1. run `forc build` 
2. add patch to the Forc.toml
3. run `forc build`
4. remove patch from the Forc.toml
5. run `forc build`

- After step 1 in the Forc.lock you should be seeing the master branch of the sway repo (for std)
- After step 3 in the Forc.lock you should be seeing the kayagokalp/1620 branch of the sway repo (for std)
- After step 5 in the Forc.lock you should be seeing the master branch of the sway repo (for std)

You can use the Forc.toml provided below.

```
[project]
authors = ["user"]
entry = "main.sw"
organization = "Fuel_Labs"
license = "Apache-2.0"
name = "parent"
implicit-std = false

[dependencies]
std = { git = "https://github.com/fuellabs/sway", branch = "master"} 
```

and to add patch in step2, you can use:

```
[patch.'https://github.com/fuellabs/sway']
std = { git = "https://github.com/fuellabs/sway", branch = "kayagokalp/1620"} 
```

or if you would like to try patching with a path:

```
[patch.'https://github.com/fuellabs/sway']
std = { path = "/path/to/std" } 
```

kayagokalp commented:
Currently, if there is a `patch` supplied for a `dependency` in the manifest file, it creates a diff between the lock and the manifest file since the lock was created without (if it is previously created) the patches. So the patches are automatically applied by `apply_pkg_diff()`. I marked `Support for patching while creating the BuildPlan from the lock file.` completed in the PR explanation but if we explicitly need the patches before `apply_pkg_diff()` I can take a look to implement it like that.

adlerjohn commented on docs/src/forc/manifest_reference.md:
```diff
@@ -99,3 +101,32 @@ Note that providing the corresponding cli options (like `--print-finalized-asm`)
+implicit-std = false
+
+[dependencies]
+std = { git = "https://github.com/fuellabs/sway", branch = "master"} 
```
```suggestion
std = { git = "https://github.com/fuellabs/sway", branch = "master" }
```

adlerjohn commented on docs/src/forc/manifest_reference.md:
```diff
@@ -99,3 +101,32 @@ Note that providing the corresponding cli options (like `--print-finalized-asm`)
+std = { git = "https://github.com/fuellabs/sway", branch = "master"} 
+
+[patch.'https://github.com/fuellabs/sway']
+std = { git = "https://github.com/fuellabs/sway", branch = "test"} 
```
```suggestion
std = { git = "https://github.com/fuellabs/sway", branch = "test" }
```

adlerjohn commented on docs/src/forc/manifest_reference.md:
````diff
@@ -99,3 +101,32 @@ Note that providing the corresponding cli options (like `--print-finalized-asm`)
+
+```toml
+[patch.'https://github.com/fuellabs/sway']
+std = { path = "/path/to/local_std_version"} 
````
```suggestion
std = { path = "/path/to/local_std_version" }
```

adlerjohn commented on forc-pkg/src/pkg.rs:
```diff
@@ -1034,6 +1069,7 @@ fn dep_to_source(pkg_path: &Path, dep: &Dependency) -> Result<Source> {
                             either `branch`, `tag` or `rev`"
                     ),
                 };
+                //We don't have any patch for this dependency
```
```suggestion
                // We don't have any patch for this dependency
```

kayagokalp commented:
Realized a bug when the lock file is removed will be taken care of shortly in the meantime converting this to a draft.

kayagokalp commented:
As it turns out, the bug was not with the code but with myself...

I was providing the wrong path for the patch but couldn't understand what is happening because lack of a good error message there. So I improved the error message. Also removed unused `Patch` enum. I changed it into a type declaration

adlerjohn commented on docs/src/forc/manifest_reference.md:
```diff
@@ -99,3 +101,32 @@ Note that providing the corresponding cli options (like `--print-finalized-asm`)
+implicit-std = false
+
+[dependencies]
+std = { git = "https://github.com/fuellabs/sway", branch = "master" }
```
Does patching work with implicit `std` too?

kayagokalp replied:
Yes, it works with implicit `std` too. To be more explicit and descriptive about what we are patching I created the doc like this. (i.e, explicitly added the dependency that I am patching). But I also realize that in this way the example is not in its most "default" form. 🤔

adlerjohn replied:
I would prefer two subsections, one for patching a library that's not `std` or `core`, and one specifically for patching the implicit `std`.

mitchmindtree commented on forc-pkg/src/pkg.rs:
```diff
@@ -732,6 +733,35 @@ pub(crate) fn fetch_deps(
+) -> Result<Source> {
+    let mut patches = manifest.patches();
+    match source {
+        //Check if the patch is for a git dependency.
```
```suggestion
        // Check if the patch is for a git dependency.
```

mitchmindtree commented on forc-pkg/src/pkg.rs:
```diff
@@ -732,6 +733,35 @@ pub(crate) fn fetch_deps(
+    match source {
+        //Check if the patch is for a git dependency.
+        Source::Git(git) => {
+            //Check if we got a patch for the git dependency.
```
```suggestion
            // Check if we got a patch for the git dependency.
```

mitchmindtree commented on forc-pkg/src/pkg.rs:
```diff
@@ -732,6 +733,35 @@ pub(crate) fn fetch_deps(
+        //Check if the patch is for a git dependency.
+        Source::Git(git) => {
+            //Check if we got a patch for the git dependency.
+            if let Some(source_patches) = patches.find(|patch| *patch.0 == git.repo.to_string()) {
```
Rather than iterating over all of the entries, `manifest.patch` is already a `BTreeMap` with log(n) lookup. We should use something like `manifest.patch.as_ref().and_then(|patches| patches.get(&git.repo.as_str()))` here.

mitchmindtree reviewed (changes requested):
Haven't tested but looking good - nice work! Just a few small things to address.

mitchmindtree commented:
@kayagokalp would you mind merging master? It's just been a couple days is all - then will happily approve :)

mitchmindtree commented:
@kayagokalp actually that said, this could probably also use a test. Would you mind adding a small test for this under `should_pass/forc`? It could be as simple as patching the implicit std URL with the local copy of `std` or something along these lines.

kayagokalp commented:
The introduction of #1790 seems to have an impact on this one, as there is a check for the dependency from its parent manifest file. If the dependency is coming from the patch it is not in the dependency section. I will be fixing this shortly and pushing the solution with the test I added.

mitchmindtree commented on forc-pkg/src/pkg.rs:
```diff
@@ -827,10 +828,25 @@ pub fn graph_to_path_map(
+                let patch = parent_manifest
+                    .patches()
+                    .find(|patches| patches.1.contains_key(&dep_name))
+                    .and_then(|patches| patches.1.get(&dep_name));
```
`find` and `and_then` used in succession can often be replaced with `find_map` - not important, just a tip!

kayagokalp replied:
Oh, thanks :) that looks better

mitchmindtree commented on test/src/e2e_vm_tests/test_programs/should_pass/forc/dependency_patching/.gitignore:
```diff
@@ -0,0 +1,2 @@
+out
+target
```
I think these should be covered by the top-level `.gitignore`?

kayagokalp replied:
yep, couldn't decide what was our general approach here while adding tests under e2e, deleting it 👍

mitchmindtree reviewed (commented):
Nice one kaya, that test is perfect! Almost there

mitchmindtree reviewed (approved):
Nice one @kayagokalp, this will be super handy :rocket:

kayagokalp commented:
There was a conflict so I had the merge manually, once again I am asking for your review 😄

adlerjohn reviewed (commented):
There should be a `test.toml` file I think: https://github.com/FuelLabs/sway/blob/bdfb7449cac1e80ee4258abf1882c3caec99111c/test/src/e2e_vm_tests/test_programs/README.md#testtoml

kayagokalp commented:
Oh sorry, added that locally but forget to include it in the commit. Adding rn.

kayagokalp commented on test/src/e2e_vm_tests/test_programs/should_pass/forc/dependency_patching/test.toml:
```diff
@@ -0,0 +1,3 @@
+category = "run"
+expected_result = { action = "return", value = 0 }
+validate_abi = true
```
Hmm, that's correct. Maybe just providing `category = "compile"` would be enough for the test.toml.

adlerjohn reviewed (approved):
utACK
